### ISSUE TITLE
Is it possible to create dataset with specified name?

### ISSUE THREAD

**Issue by RytisLT** [openebs/zfs-localpv#407]
tldr; Can I create some set of Kubernetes Resources that would cause a creation of a `Dataset`  with custom name?

If I use `PersistenceVolumeClaim` + `StorageClass` to dynamically provision a `PersistentVolume`, the newly created `PersistentVolume` is named `pvc-{uuid}` and a corresponding `Dataset` is created in the `ZPool`. As far as I understand the naming convention cannot be changed, because it is something Kubernetes internals rely on. 

I want to provide a Dataset name myself so I tried to manually create `PersistentVolume` matching `ZFSVolume` and a `PersistentVolumeClaim` that binds to the `PersistentVolume`. Yet this doesn't create the `Dataset` and `Pod` fails on startup with `rpc error: code = Internal desc = zfs get mountpoint failed, cannot open 'mypool/my-ds': dataset does not exist`

**Comment by pawanpraka1** [comment]
> I want to provide a Dataset name myself so I tried to manually create PersistentVolume matching ZFSVolume and a PersistentVolumeClaim that binds to the PersistentVolume. Yet this doesn't create the Dataset and Pod fails on startup with rpc error: code = Internal desc = zfs get mountpoint failed, cannot open 'mypool/my-ds': dataset does not exist

@RytisLT it seems like you are creating the ZFSVolume CR with status as Ready. Can you use status as Pending and then try?

**Comment by RytisLT** [comment]
I've omitted the status altogether in my first attempt. I've re-created all Resources and used Status Pending. I've got same error and ZV stays in Pending

**Comment by pawanpraka1** [comment]
can you shared the ZFSVolume CR, pvc and pv yamls?

**Comment by RytisLT** [comment]
sure thing:

edit: added namespace to pod.yaml

    #pv.yaml
    ---
    apiVersion: v1
    kind: PersistentVolume
    metadata:
      finalizers:
      - kubernetes.io/pv-protection
      name: task-pv-volume
    spec:
      accessModes:
      - ReadWriteOnce
      capacity:
        storage: 5Gi
      csi:
        driver: zfs.csi.openebs.io
        fsType: zfs
        volumeAttributes:
          openebs.io/cas-type: localpv-zfs
          openebs.io/poolname: lake/sink
          storage.kubernetes.io/csiProvisionerIdentity: 1647886600435-8081-zfs.csi.openebs.io
        volumeHandle: task-pv-volume
      claimRef:
        apiVersion: v1
        kind: PersistentVolumeClaim
        name: task-pv-claim
        namespace: test
      nodeAffinity:
        required:
          nodeSelectorTerms:
          - matchExpressions:
            - key: openebs.io/nodeid
              operator: In
              values:
              - ix-truenas
      persistentVolumeReclaimPolicy: Delete
      storageClassName: sink
      volumeMode: Filesystem
    #zv.yaml
    ---
    apiVersion: zfs.openebs.io/v1
    kind: ZFSVolume
    metadata:
      finalizers:
      - zfs.openebs.io/finalizer
      labels:
        kubernetes.io/nodename: ix-truenas
      name: task-pv-volume
      namespace: openebs
    spec:
      capacity: "5368709120"
      fsType: zfs
      ownerNodeID: ix-truenas
      poolName: lake/sink
      shared: "yes"
      volumeType: DATASET
    status:
      state: Pending
    #claim.yaml
    ---
    apiVersion: v1
    kind: PersistentVolumeClaim
    metadata:
      name: task-pv-claim
      namespace: test
    spec:
      storageClassName: sink
      accessModes:
        - ReadWriteOnce
      resources:
        requests:
          storage: 5Gi
    #pod.yaml
    ---
    apiVersion: v1
    kind: Pod
    metadata:
      name: task-pv-pod
      namespace: test
      labels:
        app.kubernetes.io/name: test-pod
    spec:
      volumes:
        - name: task-pv-storage
          persistentVolumeClaim:
            claimName: task-pv-claim
      containers:
        - name: task-pv-container
          image: nginx
          ports:
            - containerPort: 80
              name: "http-server"
          volumeMounts:
            - mountPath: "/usr/share/nginx/html"
              name: task-pv-storage

**Comment by pawanpraka1** [comment]
@RytisLT Perfect. Can you remove the finalizer from the ZFSVolume CR and then try.
```
  finalizers:
  - zfs.openebs.io/finalizer
 ```

**Comment by RytisLT** [comment]
🥳🥳🥳🥳🥳🥳🥳
That did it!! Thanks!

**Comment by RytisLT** [comment]
BTW the finalizer was added automatically after the volume transitioned to Ready state

**Comment by pawanpraka1** [comment]
>BTW the finalizer was added automatically after the volume transitioned to Ready state

Yes, that is why it should not be added before. Here since finalizer was already there, the driver will not attempt to create the volume. Removing that while applying the yaml fixed the issue and the driver will go ahead and add the finalizer once the volume is created..

**Comment by pawanpraka1** [comment]
@RytisLT would you like to mention your use case in our adopters(https://github.com/openebs/openebs/issues/2719) file.

**Comment by RytisLT** [comment]
sure thing!

**Comment by RytisLT** [comment]
@pawanpraka1 one more related question, If I want to prevent deletion of `Dataset` when `ZFSVolume` is deleted do I just simply remove the finalizers from `ZV` or is there another mechanism to prevent that?

**Comment by pawanpraka1** [comment]
@RytisLT you can add your own finalizer while creating the ZFSVolume CR.
